Add get_request_kwargs to check before requesting

kumarstack55 opened the pull request:
When we use mechanicalsoup, we sometimes want to verify a request before submitting it.

If you merge this pull request, the package will be able to provide a way for the package's users to review the request.

This is my first pull request for this project. Please let me know if I'm missing anything.

hemberger commented on mechanicalsoup/browser.py:
```diff
@@ -240,7 +240,14 @@ def __bool__(self):
 
-        return self.session.request(method, url, files=files, **kwargs)
+        request_kwargs = {"method": method, "url": url, "files": files}
+        request_kwargs.update(kwargs)
```
If the user provides `method` or `url` arguments to `**kwargs` in our API, it would currently raise a `TypeError` due to a repeated argument passed to `session.request` (I'm working on a unit test to demonstrate this now). However, in this patch, those keys would be overwritten by the `request_kwargs` dict update. It seems like it would be a little clunky to manually check for those keys, but maybe that's what's needed here to retain the expected behavior?

kumarstack55 replied:
I hadn't thought about TypeError. Thank you for pointing that out.

It looks like we can check the key for backwards compatibility. There seems to be more than one way to implement it.

Which of the following implementation methods do you prefer?

(a) Raise TypeError in the package
(b) Or create a class method, e.g. `_get_request_kwargs()` that only merges the kwargs so that Python raises a TypeError.

kumarstack55 replied:
Since we have difficulty thinking about the msg of TypeError(msg), I thought it might be more desirable for Python to raise a TypeError.

kumarstack55 replied:
536ce4a398c88a3d4bdf26da19502619ebb2de67 Resolve this issue.

hemberger reviewed (changes requested):
Hi, and thanks for your contribution to MechanicalSoup!

In principle, I think this is a fine change. I've recommended one small spelling error and identified a corner case that probably needs to be addressed. Once those are fixed, I will be ready to approve this. Then if the other maintainer approves as well we can merge it in. :)

kumarstack55 commented on mechanicalsoup/browser.py:
```diff
@@ -139,8 +139,8 @@ def post(self, *args, **kwargs):
-    def _request(self, form, url=None, **kwargs):
-        """Extract input data from the form to pass to a Requests session."""
+    @classmethod
+    def get_reuqest_kwargs(cls, form, url=None, **kwargs):
```
Thank you for pointing this out. It has been corrected.

kumarstack55 commented:
The group of jobs in travisci are listed as Queued, but when I click on Details to check, it looks like all the jobs are successful.

kumarstack55 commented:
> The group of jobs in travisci are listed as Queued, but when I click on Details to check, it looks like all the jobs are successful.

@hemberger Is there anything else I should do? :)

moy commented:
Commits should be squashed and the typo should be fixed in the commits summary (I just fixed it in the PR title). We keep a clean history hence a PR should include several commits only if the change needs to be split into several commits, not as leftover of previous mistakes.

moy commented on mechanicalsoup/browser.py:
```diff
@@ -139,8 +139,16 @@ def post(self, *args, **kwargs):
-    def _request(self, form, url=None, **kwargs):
-        """Extract input data from the form to pass to a Requests session."""
+    @classmethod
+    def _get_request_kwargs(cls, method, url, **kwargs):
```
`cls` is unused here, right? Why not use `@staticmethod` instead of `@classmethod` then?

kumarstack55 replied:
Thanks for the good point. I'll use `@staticmethod` to avoid making `cls` useless.

moy commented on mechanicalsoup/browser.py:
```diff
@@ -139,8 +139,16 @@ def post(self, *args, **kwargs):
+    @classmethod
+    def _get_request_kwargs(cls, method, url, **kwargs):
+        # This method exists to raise a TypeError when a method or url is
+        # specified in the kwargs.
```
Shouldn't this comment be a docstring?

kumarstack55 replied:
Actually, this is my first time writing a docstring.

I'll be referring to other methods to write docstrings. Please let me know if you have any concerns.

moy commented on tests/test_browser.py:
```diff
@@ -38,6 +38,48 @@ def test_submit_online(httpbin):
+    try:
+        browser.get_request_kwargs(form, page.url, **kwargs)
+    except TypeError:
+        type_error_raised = True
```
Please use [pytest.raises](https://docs.pytest.org/en/stable/reference.html?highlight=raises#pytest.raises) to assert that an exception is raised.

kumarstack55 replied:
Thanks, I'll rewrite it using pytest.raises.

moy commented on tests/test_browser.py:
```diff
@@ -38,6 +38,48 @@ def test_submit_online(httpbin):
+        # pylint: disable=redundant-keyword-arg
+        browser.get_request_kwargs(form, page.url, **kwargs)
+    except TypeError:
+        type_error_raised = True
```
Likewise.

kumarstack55 replied:
Thanks, I'll rewrite it using pytest.raises.

moy reviewed (changes requested):
On overall, this looks good, but a few more nits to be fixed before this can be merged.

kumarstack55 commented:
Originally, this pull request was intended to implement `_get_request_kwargs()`, but the name was changed and `get_request_kwargs()` is the current name.

Therefore, please let me change the title of the Pull Request.

kumarstack55 commented:
> Commits should be squashed and the typo should be fixed in the commits summary (I just fixed it in the PR title).

I squashed the commits to one. :sparkles:

moy commented:
On overall this looks good to me. The commit message would deserve a bit more love I think (explain why the change is needed and good).

kumarstack55 commented:
> On overall this looks good to me. The commit message would deserve a bit more love I think (explain why the change is needed and good).

Let me make the commit log more understandable and lovely. I don't plan to make any changes to the code at this time.

kumarstack55 commented:
@hemberger @moy Let me know if there's anything else I need to do to get the pull request merged.

hemberger commented on mechanicalsoup/browser.py:
```diff
@@ -139,8 +139,17 @@ def post(self, *args, **kwargs):
+        return request_kwargs
+
+    @classmethod
+    def get_request_kwargs(cls, form, url=None, **kwargs):
```
Since this is a public method, it will need a docstring.

moy replied:
Hmm, and this should be a @staticmethod, not @classmethod, right?

hemberger replied:
I think they used classmethod because it calls
```python
return cls._get_request_kwargs(method, url, files=files, **kwargs)
```
However, this could be `Browser._get_request_kwargs`. I'm happy with it changed or as-is.

moy replied:
OK, I'm fine with both too.

kumarstack55 replied:
I added a simple docstring to the public method.

hemberger reviewed (changes requested):
Overall, the changes look great. I've identified one simple change I'd like to see, but then I think this is ready to be merged. Thanks!

hemberger commented:
Thanks again for your contribution!

kumarstack55 commented:
Thank you for merging. :smile:

By the way, it occurs to me that this was the first time in my life that I've ever had my code reviewed.
I don't have a job writing programs. Furthermore, I had never even written a docstring before.
(I do some scraping as a hobby, so having Mechanicalsoup available would be very helpful. )

For me, all of the points you @hemberger @moy have given me are excellent. :sparkles:
I'm sure it would be very difficult for you to review my code.
Thanks to your thoughtful comments, I managed to get it merged. 

Again, thank you for the series of responses and for merging my request. :+1:

moy commented:
Thanks for the feedback. In another life, I'm a teacher, and that's what I try to explain my students: code review is not "I went through it and it doesn't look too bad", but can be much more picky *and constructive* than that ;-).
